src/main/java/org/broadinstitute/hellbender/engine/dataflow/transforms/composite/AddContextDataToReadOptimizedFn.java

```diff
@@ -0,0 +1,280 @@
+                throw new RuntimeException("Sorry, Hadoop paths aren't yet supported");
+            } else {
+                // read from local file (this only makes sense if every worker sees the same thing, e.g. if we're running locally)
+                reader = SamReaderFactory.make().validationStringency(ValidationStringency.SILENT).open(new File(bam));
+            }
+            query = reader.queryOverlapping(shard.interval.getContig(), shard.interval.getStart(), shard.interval.getEnd());
+
+        }
+
+        @Override
+        public boolean hasNext() {
+            if (null==nextOutput) nextOutput = tryNext();
+            return (null!=nextOutput);
+        }
+
+        @Override
+        public ContextShard next() {
+            if (null==nextOutput) nextOutput = tryNext();
+            if (null==nextOutput) throw new NoSuchElementException();
+            ContextShard ret =  nextOutput;
+            nextOutput = null;
+            return ret;
+        }
+
+        // returns the next shard, if any. Otherwise, closes the reader and returns null.
+        private ContextShard tryNext() {
+            if (readerClosed) return null;
+            ContextShard ret = null;
+            while (query.hasNext()) {
+
+                SAMRecord r = query.next();
+                SAMRecordToGATKReadAdapter g = new SAMRecordToGATKReadAdapter(r);
+                // yes, it'd be a tad faster to check before the wrapping.
+                // But this keeps the code a tad simpler.
+                if (!accept(g, shard.interval)) continue;
+                if (null!=optFilter) {
+                    // skip reads that don't pass the filter
+                    if (!optFilter.test(g)) continue;
+                }
+                if (!g.isUnmapped()) {
+                    // error out if we accept a read that sticks out too far
+                    // (margin was too tight, the shard may end up missing relevant variants)
+                    if (r.getAlignmentEnd()>lastValidPos) {
+                        throw new GATKException("Margin was too tight, a read sticks out by "+(r.getAlignmentEnd()-shard.interval.getEnd())+", going all the way to "+r.getAlignmentEnd());
+                    }
+                    if (r.getAlignmentStart()<firstValidPos) {
+                        throw new GATKException("Margin was too tight, a read starts early by "+(shard.interval.getStart()-r.getAlignmentStart())+", starting at "+r.getAlignmentStart());
+                    }
+                }
+
+                while (currentSubShard.getEnd() < r.getStart()) {
+                    if (!readsSoFar.isEmpty()) {
+                        // ship this one.
+                        ret = shard.split(currentSubShard).withReads(readsSoFar);
+                        readsSoFar = new ArrayList<>();
+                    }
+                    // move to the next shard
+                    currentSubShard = subshards.get(++currentSubShardIndex);
```

Comment:
That'll teach me to answer questions but not put them in a comment in the code. 
There is indeed no way we could go out of bounds here because all the reads overlap the shard. Therefore if the "while" condition is true (we have a read starting after the current subshard) then we are not at the last subshard yet.